virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,278 @@ func (svc *googleService) insertSSHKey() error {
 
 	return svc.waitForOperationReady(op.Name)
 }
+
+// getNetworkList gets the list of networks from the given project
+func (svc *googleService) getNetworkList() ([]*googlecloud.Network, error) {
```

Comment:
In general, what is the decision filter to choose googleService or VM struct type pointer receiver for any method?
Also, why there are many duplicate methods on these? what's the difference.
e.g. GetMachineTypeList in utils.go and vm.go

Reply:
The way GCP actions are handled in Libretto is that all the actions are first received on VM struct as a receiver in vm.go. It does the pre-processing if any and then calls corresponding function on googleService receiver in util.go. This is how it was done for all the existing actions such as provision, start, halt, etc.

All the functions in vm.go are received on VM struct and all the functions (except getService) in util.go are received on googleService. The functions in vm.go do not interact with GCP SDK directly. All the interaction with GCP SDK happens via util.go. That's why you will se same function (e.g. provision or getMachineTypeList) defined with VM as well as googleService as receivers in vm.go and util.go respectively. Function in util.go are not exposed. They can be invoked only by functions in the same package. I have followed the same practice with new functions. Hope this helps. Let me know in case of any question.

Reply:
Thanks!